File: dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/Application/Authentication/Bot/BotAuthenticationBase.cs
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;
using Microsoft.Teams.AI.Exceptions;
using Microsoft.Teams.AI.State;
using Newtonsoft.Json.Linq;

namespace Microsoft.Teams.AI
{
    /// <summary>
    /// Base class for bot authentication that handles common logic
    /// </summary>
    internal abstract class BotAuthenticationBase<TState>
        where TState : TurnState, new()
    {
        /// <summary>
        /// Name of the authentication handler
        /// </summary>
        protected string _name;

        /// <summary>
        /// Storage to save turn state
        /// </summary>
        protected IStorage _storage;

        /// <summary>
        /// Callback when user sign in success
        /// </summary>
        internal Func<ITurnContext, TState, Task>? _userSignInSuccessHandler;

        /// <summary>
        /// Callback when user sign in fail
        /// </summary>
        internal Func<ITurnContext, TState, AuthException, Task>? _userSignInFailureHandler;

        /// <summary>
        /// Initializes the class
        /// </summary>
        /// <param name="app">The application instance</param>
        /// <param name="name">The name of authentication handler</param>
        /// <param name="storage">The storage to save turn state</param>
        public BotAuthenticationBase(Application<TState> app, string name, IStorage? storage = null)
        {
            _name = name;
            _storage = storage ?? new MemoryStorage();

            // Add application routes to handle OAuth callbacks
            app.AddRoute(this.VerifyStateRouteSelector, async (context, state, cancellationToken) =>
            {
                await this.HandleSignInActivity(context, state, cancellationToken);
            }, true);

            app.AddRoute(this.TokenExchangeRouteSelector, async (context, state, cancellationToken) =>
            {
                await this.HandleSignInActivity(context, state, cancellationToken);
            }, true);
        }

        /// <summary>
        /// Authenticate current user
        /// </summary>
        /// <param name="context">The turn context</param>
        /// <param name="state">The turn state</param>
        /// <param name="cancellationToken">The cancellation token</param>
        /// <returns>The sign in response</returns>
        public async Task<string?> AuthenticateAsync(ITurnContext context, TState state, CancellationToken cancellationToken = default)
        {
            // Get property names to use
            string userAuthStatePropertyName = GetUserAuthStatePropertyName(context);
            string userDialogStatePropertyName = GetUserDialogStatePropertyName(context);

            // Save message if not signed in
            if (!TryGetUserAuthState(context, state, out _))
            {
                state.Conversation.Add(userAuthStatePropertyName, new Dictionary<string, string>()
                {
                    {"message", context.Activity.Text }
                });
            }

            DialogTurnResult result = await RunDialog(context, state, userDialogStatePropertyName, cancellationToken);
            if (result.Status == DialogTurnStatus.Complete)
            {
                // Delete user auth state
                DeleteAuthFlowState(context, state);
                TokenResponse? tokenResponse = result.Result as TokenResponse;
                if (tokenResponse == null)
                {
                    // Completed dialog without a token.
                    // This could mean the user declined the consent prompt in the previous turn.
                    // Retry authentication flow again.
                    return await AuthenticateAsync(context, state, cancellationToken);
                }
                else
                {
                    // Return token
                    return tokenResponse.Token;
                }
            }

            return null;
        }

        /// <summary>
        /// Whether the current activity is a valid activity that supports authentication
        /// </summary>
        /// <param name="context">The turn context</param>
        /// <returns>True if valid. Otherwise, false.</returns>
        public virtual bool IsValidActivity(ITurnContext context)
        {
            return context.Activity.Type == ActivityTypes.Message
                && !string.IsNullOrEmpty(context.Activity.Text);
        }

        /// <summary>
        /// Handles the signin/verifyState activity. The onUserSignInSuccess and onUserSignInFailure handlers will be called based on the result.
        /// </summary>
        /// <param name="context">The turn context</param>
        /// <param name="state">The turn state</param>
        /// <param name="cancellationToken">The cancellation token</param>
        /// <returns></returns>
        public async Task HandleSignInActivity(ITurnContext context, TState state, CancellationToken cancellationToken)
        {
            try
            {
                string userDialogStatePropertyName = GetUserDialogStatePropertyName(context);
                DialogTurnResult result = await ContinueDialog(context, state, userDialogStatePropertyName, cancellationToken);

                if (result.Status == DialogTurnStatus.Complete)
                {
                    // OAuthPrompt dialog should have sent an invoke response already.
                    TokenResponse? tokenResponse = result.Result as TokenResponse;
                    if (tokenResponse != null)
                    {
                        // Successful sign in
                        AuthUtilities.SetTokenInState(state, _name, tokenResponse.Token);

                        if (TryGetUserAuthState(context, state, out Dictionary<string, string> userAuthState))
                        {
                            if (userAuthState.ContainsKey("message"))
                            {
                                context.Activity.Text = userAuthState["message"];
                            }
                            else
                            {
                                context.Activity.Text = "";
                            }
                        }
                        else
                        {
                            context.Activity.Text = "";
                        }

                        if (_userSignInSuccessHandler != null)
                        {
                            await _userSignInSuccessHandler(context, state);
                        }
                    }
                    else
                    {
                        // Failed sign in
                        if (_userSignInFailureHandler != null)
                        {
                            await _userSignInFailureHandler(context, state, new AuthException("Authentication flow completed without a token.", AuthExceptionReason.CompletionWithoutToken));
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                string message = $"Unexpected error encountered while signing in: {ex.Message}.\nIncoming activity details: type: {context.Activity.Type}, name: {context.Activity.Name}";

                if (_userSignInFailureHandler != null)
                {
                    await _userSignInFailureHandler(context, state, new AuthException(message));
                }
            }
        }

        /// <summary>
        /// The handler function is called when the user has successfully signed in
        /// </summary>
        /// <param name="handler">The handler function to call when the user has successfully signed in</param>
        public void OnUserSignInSuccess(Func<ITurnContext, TState, Task> handler)
        {
            _userSignInSuccessHandler = handler;
        }

        /// <summary>
        /// The handler function is called when the user sign in flow fails
        /// </summary>
        /// <param name="handler">The handler function to call when the user failed to signed in</param>
        public void OnUserSignInFailure(Func<ITurnContext, TState, AuthException, Task> handler)
        {
            _userSignInFailureHandler = handler;
        }

        /// <summary>
        /// The route selector for signin/verifyState activity
        /// </summary>
        /// <param name="context">The turn context</param>
        /// <param name="cancellationToken">The cancellation token</param>
        /// <returns>True if the activity should be handled by current authentication handler. Otherwise, false.</returns>
        internal virtual Task<bool> VerifyStateRouteSelector(ITurnContext context, CancellationToken cancellationToken)
        {
            string? settingName = GetSettingNameFromContextActivityValue(context);

            return Task.FromResult(
                context.Activity.Type == ActivityTypes.Invoke &&
                context.Activity.Name == SignInConstants.VerifyStateOperationName &&
                this._name == settingName
            );
        }

        /// <summary>
        /// The route selector for signin/tokenExchange activity
        /// </summary>
        /// <param name="context">The turn context</param>
        /// <param name="cancellationToken">The cancellation token</param>
        /// <returns>True if the activity should be handled by current authentication handler. Otherwise, false.</returns>
        internal virtual Task<bool> TokenExchangeRouteSelector(ITurnContext context, CancellationToken cancellationToken)
        {
            string? settingName = GetSettingNameFromContextActivityValue(context);

            return Task.FromResult(context.Activity.Type == ActivityTypes.Invoke &&
                context.Activity.Name == SignInConstants.TokenExchangeOperationName &&
                this._name == settingName
            );
        }

        static internal void SetSettingNameInContextActivityValue(ITurnContext context, string settingName)
        {
            JObject value;
            if (context.Activity.Value != null)
            {
                value = JObject.FromObject(context.Activity.Value);
            }
            else
            {
                value = new();
            }

            value.Add("settingName", settingName);
            context.Activity.Value = value;

            return;
        }

        static internal string? GetSettingNameFromContextActivityValue(ITurnContext context)
        {
            if (context.Activity.Value == null)
            {
                return null;
            }

            JObject value = JObject.FromObject(context.Activity.Value);
            JToken? settingName = value.GetValue("settingName");

            return settingName != null ? settingName.ToString() : null;
        }

        private string GetUserAuthStatePropertyName(ITurnContext context)
        {
            return $"__{context.Activity.From.Id}:{_name}:Bot:AuthState__";
        }

        private string GetUserDialogStatePropertyName(ITurnContext context)
        {
            return $"__{context.Activity.From.Id}:{_name}:DialogState__";
        }

        private bool TryGetUserAuthState(ITurnContext context, TState state, out Dictionary<string, string> authState)
        {
            string propertyName = GetUserAuthStatePropertyName(context);
            return state.Conversation.TryGetValue(propertyName, out authState);
        }

        public void DeleteAuthFlowState(ITurnContext context, TState state)
        {
            // Delete user auth state
            string userAuthStatePropertyName = GetUserAuthStatePropertyName(context);
            if (state.Conversation.ContainsKey(userAuthStatePropertyName))
            {
                state.Conversation.Remove(userAuthStatePropertyName);
            }

            // Delete user dialog state
            string userDialogStatePropertyName = GetUserDialogStatePropertyName(context);
            if (state.Conversation.ContainsKey(userDialogStatePropertyName))
            {
                state.Conversation.Remove(userDialogStatePropertyName);
            }
        }

        /// <summary>
        /// Run or continue the authentication dialog.
        /// </summary>
        /// <param name="context">The turn context</param>
        /// <param name="state">The turn state</param>
        /// <param name="dialogStateProperty">The property name for storing dialog state.</param>
        /// <param name="cancellationToken">The cancellation token</param>
        /// <returns>Dialog turn result that contains token if sign in success</returns>
        public abstract Task<DialogTurnResult> RunDialog(ITurnContext context, TState state, string dialogStateProperty, CancellationToken cancellationToken = default);

        /// <summary>
        /// Continue the authentication dialog.
        /// </summary>
        /// <param name="context">The turn context</param>
        /// <param name="state">The turn state</param>
        /// <param name="dialogStateProperty">The property name for storing dialog state.</param>
        /// <param name="cancellationToken">The cancellation token</param>
        /// <returns>Dialog turn result that contains token if sign in success</returns>
        public abstract Task<DialogTurnResult> ContinueDialog(ITurnContext context, TState state, string dialogStateProperty, CancellationToken cancellationToken = default);
    }
}

File: js/packages/teams-ai/src/authentication/BotAuthenticationBase.ts
import { DialogState, DialogTurnResult, DialogTurnStatus } from 'botbuilder-dialogs';
import { TurnState } from '../TurnState';
import { Application } from '../Application';
import {
    ActivityTypes,
    MemoryStorage,
    TurnContext,
    Storage,
    verifyStateOperationName,
    tokenExchangeOperationName,
    TokenResponse
} from 'botbuilder';
import { AuthError } from './Authentication';

/**
 * @internal
 */
interface UserAuthState {
    message?: string;
}

/**
 * @private
 */
const IS_SIGNED_IN_KEY = '__InSignInFlow__';

/**
 * @internal
 * Base class to handle Teams conversational bot authentication.
 * @template TState - The type of the turn state.
 */
export abstract class BotAuthenticationBase<TState extends TurnState> {
    protected _storage: Storage;
    protected _settingName: string;
    private _userSignInSuccessHandler?: (context: TurnContext, state: TState) => Promise<void>;
    private _userSignInFailureHandler?: (context: TurnContext, state: TState, error: AuthError) => Promise<void>;

    /**
     * Creates a new instance of BotAuthenticationBase.
     * @param {Application<TState>} app - The application instance.
     * @param {string} settingName - The name of the setting.
     * @param {Storage} [storage] - The storage to save states.
     */
    public constructor(app: Application<TState>, settingName: string, storage?: Storage) {
        this._settingName = settingName;

        this._storage = storage || new MemoryStorage();

        // Add application routes to handle OAuth callbacks
        app.addRoute(
            this.verifyStateRouteSelector.bind(this),
            async (context, state) => {
                await this.handleSignInActivity(context, state);
            },
            true
        );
        app.addRoute(
            this.tokenExchangeRouteSelector.bind(this),
            async (context, state) => {
                await this.handleSignInActivity(context, state);
            },
            true
        );
    }

    /**
     * Authenticates the user.
     * @param {TurnContext} context - The turn context.
     * @param {TState} state - The turn state.
     * @returns {Promise<string | undefined>} - The authentication token, or undefined if authentication failed.
     */
    public async authenticate(context: TurnContext, state: TState): Promise<string | undefined> {
        // Get property names to use
        const userAuthStatePropertyName = this.getUserAuthStatePropertyName(context);
        const userDialogStatePropertyName = this.getUserDialogStatePropertyName(context);

        // Save message if not signed in
        if (!this.getUserAuthState(context, state)) {
            (state.conversation as any)[userAuthStatePropertyName] = {
                message: context.activity.text
            };
        }

        const results = await this.runDialog(context, state, userDialogStatePropertyName);
        if (results.status === DialogTurnStatus.complete) {
            // Delete user auth state
            this.deleteAuthFlowState(context, state);

            if (results.result?.token) {
                // Return token
                return results.result?.token;
            } else {
                // Completed dialog without a token.
                // This could mean the user declined the consent prompt in the previous turn.
                // Retry authentication flow again.
                return await this.authenticate(context, state);
            }
        }

        return undefined;
    }

    /**
     * Checks if the activity is a valid message activity
     * @param {TurnContext} context - The turn context.
     * @returns {boolean} - True if the activity is a valid message activity.
     */
    public isValidActivity(context: TurnContext): boolean {
        // Should be a message activity with non-empty text property.
        return (
            context.activity.type === ActivityTypes.Message &&
            context.activity.text != null &&
            context.activity.text != undefined &&
            context.activity.text.length > 0
        );
    }

    /**
     * The handler function is called when the user has successfully signed in
     * @template TState
     * @param {(context: TurnContext, state: TState) => Promise<void>} handler The handler function to call when the user has successfully signed in
     */
    public onUserSignInSuccess(handler: (context: TurnContext, state: TState) => Promise<void>): void {
        this._userSignInSuccessHandler = handler;
    }

    /**
     * The handler function is called when the user sign in flow fails
     * @template TState
     * @param {(context: TurnContext, state: TState) => Promise<void>} handler The handler function to call when the user failed to signed in
     */
    public onUserSignInFailure(
        handler: (context: TurnContext, state: TState, error: AuthError) => Promise<void>
    ): void {
        this._userSignInFailureHandler = handler;
    }

    /**
     * Handles the signin/verifyState activity. The onUserSignInSuccess and onUserSignInFailure handlers will be called based on the result.
     * @param {TurnContext} context - The turn context.
     * @param {TState} state - The turn state.
     */
    public async handleSignInActivity(context: TurnContext, state: TState): Promise<void> {
        try {
            const userDialogStatePropertyName = this.getUserDialogStatePropertyName(context);
            const result = await this.continueDialog(context, state, userDialogStatePropertyName);

            if (result.status === DialogTurnStatus.complete) {
                // OAuthPrompt dialog should have sent an invoke response already.

                if (result.result?.token) {
                    // Successful sign in
                    setTokenInState(state, this._settingName, result.result.token);

                    // Get user auth state
                    const userAuthState = this.getUserAuthState(context, state);

                    // Restore previous user message
                    context.activity.text = userAuthState.message || '';

                    await this._userSignInSuccessHandler?.(context, state);
                } else {
                    // Failed sign in
                    await this._userSignInFailureHandler?.(
                        context,
                        state,
                        new AuthError('Authentication flow completed without a token.', 'completionWithoutToken')
                    );
                }
            }
        } catch (e) {
            const errorMessage = e instanceof Error ? e.message : JSON.stringify(e);
            const message = `Unexpected error encountered while signing in: ${errorMessage}.
                Incoming activity details: type: ${context.activity.type}, name: ${context.activity.name}`;

            await this._userSignInFailureHandler?.(context, state, new AuthError(message));
        }
    }

    /**
     * Deletes the user auth state and user dialog state from the turn state. So that the next message can start a new authentication flow.
     * @param {TurnContext} context - The turn context.
     * @param {TState} state - The turn state.
     */
    public deleteAuthFlowState(context: TurnContext, state: TState) {
        // Delete user auth state
        const userAuthStatePropertyName = this.getUserAuthStatePropertyName(context);
        if (this.getUserAuthState(context, state)) {
            delete (state.conversation as any)[userAuthStatePropertyName];
        }

        // Delete user dialog state
        const userDialogStatePropertyName = this.getUserDialogStatePropertyName(context);
        if (this.getUserDialogState(context, state)) {
            delete (state.conversation as any)[userDialogStatePropertyName];
        }
    }

    /**
     * Gets the property name for storing user authentication state.
     * @param {TurnContext} context - The turn context.
     * @returns {string} - The property name.
     */
    public getUserAuthStatePropertyName(context: TurnContext): string {
        return `__${context.activity.from.id}:${this._settingName}:Bot:AuthState__`;
    }

    /**
     * Gets the property name for storing user dialog state.
     * @param {TurnContext} context - The turn context.
     * @returns {string} - The property name.
     */
    public getUserDialogStatePropertyName(context: TurnContext): string {
        return `__${context.activity.from.id}:${this._settingName}:DialogState__`;
    }

    private getUserAuthState(context: TurnContext, state: TState): UserAuthState {
        return (state.conversation as any)[this.getUserAuthStatePropertyName(context)] as UserAuthState;
    }

    private getUserDialogState(context: TurnContext, state: TState): DialogState {
        const userDialogStatePropertyName = this.getUserDialogStatePropertyName(context);
        return (state.conversation as any)[userDialogStatePropertyName] as DialogState;
    }

    public async verifyStateRouteSelector(context: TurnContext): Promise<boolean> {
        return (
            context.activity.type === ActivityTypes.Invoke &&
            context.activity.name === verifyStateOperationName &&
            this._settingName == context.activity.value['settingName']
        );
    }

    public async tokenExchangeRouteSelector(context: TurnContext): Promise<boolean> {
        return (
            context.activity.type === ActivityTypes.Invoke &&
            context.activity.name === tokenExchangeOperationName &&
            this._settingName == context.activity.value['settingName']
        );
    }

    /**
     * Run or continue the authentication dialog.
     * @param {TurnContext} context - The turn context.
     * @param {TState} state - The turn state.
     * @param {string} dialogStateProperty - The property name for storing dialog state.
     * @returns {Promise<DialogTurnResult<TokenResponse>>} - A promise that resolves to the dialog turn result containing the token response.
     */
    public abstract runDialog(
        context: TurnContext,
        state: TState,
        dialogStateProperty: string
    ): Promise<DialogTurnResult<TokenResponse>>;

    /**
     * Continues the authentication dialog.
     * @param {TurnContext} context - The turn context.
     * @param {TState} state - The turn state.
     * @param {string} dialogStateProperty - The property name for storing dialog state.
     * @returns {Promise<DialogTurnResult<TokenResponse>>} - A promise that resolves to the dialog turn result containing the token response.
     */
    public abstract continueDialog(
        context: TurnContext,
        state: TState,
        dialogStateProperty: string
    ): Promise<DialogTurnResult<TokenResponse>>;
}

/**
 * Sets the setting name in the context.activity.value object.
 * The setting name is needed in signIn/verifyState` and `signIn/tokenExchange` route selector to accurately route to the correct authentication setting.
 * @param {TurnContext} context The turn context object
 * @param {string} settingName The auth setting name
 */
export function setSettingNameInContextActivityValue(context: TurnContext, settingName: string) {
    if (typeof context.activity.value == 'object') {
        context.activity.value['settingName'] = settingName;
    } else {
        context.activity.value = {
            settingName: settingName
        };
    }
}

/**
 * Sets the token in the turn state
 * @param {TurnState} state The turn state
 * @param {string} settingName The name of the setting
 * @param {string} token The token to set
 * @internal
 */
export function setTokenInState<TState extends TurnState>(state: TState, settingName: string, token: string) {
    if (!state.temp.authTokens) {
        state.temp.authTokens = {};
    }

    state.temp.authTokens[settingName] = token;
}

/**
 * Deletes the token from the turn state
 * @param {TurnState} state The turn state
 * @param {string} settingName The name of the setting
 */
export function deleteTokenFromState<TState extends TurnState>(state: TState, settingName: string) {
    if (!state.temp.authTokens || !state.temp.authTokens[settingName]) {
        return;
    }

    delete state.temp.authTokens[settingName];
}

/**
 * Determines if the user is in the sign in flow.
 * @template TState
 * @param {TState} state - the turn state
 * @returns {string | undefined} The setting name if the user is in sign in flow. Otherwise, undefined.
 */
export function userInSignInFlow<TState extends TurnState>(state: TState): string | undefined {
    if (IS_SIGNED_IN_KEY in state.user && typeof state.user[IS_SIGNED_IN_KEY] == 'string') {
        // returns the connection name if the user is in the sign in flow
        return state.user[IS_SIGNED_IN_KEY];
    }

    return;
}

/**
 * Update the turn state to indicate the user is in the sign in flow by providing the authentication setting name used.
 * @template TState
 * @param {TState} state - the turn state
 * @param {string} settingName - the authentication setting name.
 */
export function setUserInSignInFlow<TState extends TurnState>(state: TState, settingName: string) {
    (state.user as any)[IS_SIGNED_IN_KEY] = settingName;
}

/**
 * Deletes the user sign in flow state from the turn state.
 * @template TState
 * Determines if the user is in the sign in flow.
 * @param {TState} state - the turn state
 */
export function deleteUserInSignInFlow<TState extends TurnState>(state: TState) {
    delete (state.user as any)[IS_SIGNED_IN_KEY];
}
